### ISSUE TITLE
Make no-self-assign allow exporting symbols

### ISSUE THREAD

**Issue by RoboErikG** [eslint/eslint#10779]
**What rule do you want to change?**

no-self-assign should ignore self assignments of the form object['name'] = object.name. This is used for exporting symbols that shouldn't be renamed by compression like in [Closure's compiler.](https://developers.google.com/closure/compiler/docs/api-tutorial3#removal):

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behavior

**Please provide some example code that this change will affect:**

https://github.com/google/blockly/blob/master/core/workspace.js#L612 is an example of where this type of self assignment gets used in Blockly.

```js
Blockly.Workspace.prototype['clear'] = Blockly.Workspace.prototype.clear;
Blockly.Workspace.prototype['clearUndo'] =
    Blockly.Workspace.prototype.clearUndo;
Blockly.Workspace.prototype['addChangeListener'] =
    Blockly.Workspace.prototype.addChangeListener;
Blockly.Workspace.prototype['removeChangeListener'] =
    Blockly.Workspace.prototype.removeChangeListener;
```

**What does the rule currently do for this code?**

  614:40  error  'Blockly.Workspace.prototype.clear' is assigned to itself                 no-self-assign
  616:5   error  'Blockly.Workspace.prototype.clearUndo' is assigned to itself             no-self-assign
  618:5   error  'Blockly.Workspace.prototype.addChangeListener' is assigned to itself     no-self-assign
  620:5   error  'Blockly.Workspace.prototype.removeChangeListener' is assigned to itself  no-self-assign

**What will the rule do after it's changed?**

No output.
